src/snapshot.rs

```diff
@@ -0,0 +1,185 @@
+use rayon::iter::{IntoParallelIterator, ParallelExtend, ParallelIterator};
+use std::convert::TryFrom;
+
+const WASM_PAGE_SIZE: u32 = 65_536;
+const NATIVE_PAGE_SIZE: u32 = 4_096;
+
+/// A "snapshot" of Wasm state from its default value after having been initialized.
+pub struct Snapshot<'a> {
+    /// Maps global index to its initialized value.
+    pub globals: Vec<wasmtime::Val>,
+
+    /// A new minimum size for each memory (in units of pages).
+    pub memory_mins: Vec<u32>,
+
+    /// Segments of non-zero memory.
+    pub data_segments: Vec<DataSegment<'a>>,
+
+    /// Snapshots for each nested instantiation.
+    pub instantiations: Vec<Snapshot<'a>>,
+}
+
+/// A data segment initializer for a memory.
+pub struct DataSegment<'a> {
+    /// The index of this data segment's memory.
+    pub memory_index: u32,
+    /// The offset within the memory that `data` should be copied to.
+    pub offset: u32,
+    /// This segment's (non-zero) data.
+    pub data: &'a [u8],
+}
+
+/// Snapshot the given instance's globals, memories, and instances from the Wasm
+/// defaults.
+//
+// TODO: when we support reference types, we will have to snapshot tables.
+pub fn snapshot<'a>(store: &'a wasmtime::Store, instance: &wasmtime::Instance) -> Snapshot<'a> {
+    log::debug!("Snapshotting the initialized state");
+
+    assert!(wasmtime::Store::same(store, &instance.store()));
+
+    let globals = snapshot_globals(instance);
+    let (memory_mins, data_segments) = snapshot_memories(instance);
+    let instantiations = snapshot_instantiations(store, instance);
+
+    Snapshot {
+        globals,
+        memory_mins,
+        data_segments,
+        instantiations,
+    }
+}
+
+/// Get the initialized values of all globals.
+fn snapshot_globals(instance: &wasmtime::Instance) -> Vec<wasmtime::Val> {
+    log::debug!("Snapshotting global values");
+    let mut globals = vec![];
+    let mut index = 0;
+    loop {
+        let name = format!("__wizer_global_{}", index);
+        match instance.get_global(&name) {
+            None => break,
+            Some(global) => {
+                globals.push(global.get());
+                index += 1;
+            }
+        }
+    }
+    globals
+}
+
+/// Find the initialized minimum page size of each memory, as well as all
+/// regions of non-zero memory.
+fn snapshot_memories<'a>(instance: &wasmtime::Instance) -> (Vec<u32>, Vec<DataSegment<'a>>) {
+    log::debug!("Snapshotting memories");
+
+    // Find and record non-zero regions of memory (in parallel).
+    let mut memory_mins = vec![];
+    let mut data_segments = vec![];
+    let mut memory_index = 0;
+    loop {
+        let name = format!("__wizer_memory_{}", memory_index);
+        match instance.get_memory(&name) {
```

Comment:
FWIW I like to model this as `let memory = match ...` to avoid the extra level of indentation for the whole `Some` case